cni-repair-controller/src/lib.rs

```diff
@@ -0,0 +1,247 @@
+const EVENT_REASON: &str = "LinkerdCNINotConfigured";
+
+#[derive(Clone, Debug)]
+pub struct Metrics {
+    queue_overflow: Counter<u64>,
+    deleted_pods: Counter<u64>,
+}
+
+pub fn run(
+    rt: &mut Runtime,
+    node_name: String,
+    controller_pod_name: String,
+    metrics: Metrics,
+) -> JoinHandle<()> {
+    let pod_evts = rt.watch_all::<Pod>(
+        watcher::Config::default()
+            .labels(DATA_PLANE_LABEL)
+            .fields(&format!("spec.nodeName={node_name}")),
+    );
+    let (tx, rx) = mpsc::channel(EVENT_CHANNEL_CAPACITY);
+    tokio::spawn(process_events(pod_evts, tx, metrics.clone()));
+
+    let client = rt.client();
+    tokio::spawn(process_pods(client, controller_pod_name, rx, metrics))
+}
+
+async fn process_events(
+    pod_evts: impl Stream<Item = watcher::Event<Pod>>,
+    tx: Sender<ObjectReference>,
+    metrics: Metrics,
+) {
+    tokio::pin!(pod_evts);
+    while let Some(evt) = pod_evts.next().await {
+        tracing::trace!(?evt);
+        if let watcher::Event::Applied(pod) = evt {
+            let status = if let Some(ref status) = pod.status {
+                status.clone()
+            } else {
+                tracing::info!("Skipped, no status");
+                continue;
+            };
+
+            let terminated = status
+                .init_container_statuses
+                .unwrap_or_default()
+                .iter()
+                .find(|status| status.name == "linkerd-network-validator")
+                .and_then(|status| status.last_state.as_ref())
+                .and_then(|status| status.terminated.as_ref())
+                .map(|term| term.exit_code == UNSUCCESSFUL_EXIT_CODE)
+                .unwrap_or(false);
+
+            let deleting = pod.metadata.deletion_timestamp.is_some();
+
+            if terminated && !deleting {
+                let namespace = pod.namespace().unwrap();
+                let name = pod.name_any();
+                let object_ref = pod.object_ref(&());
+                // this avoids blocking the event loop
+                match tx.try_send(object_ref) {
+                    Ok(_) => {}
+                    Err(TrySendError::Full(_)) => {
+                        tracing::warn!(%namespace, %name, "Dropped event (channel full)");
+                        metrics.queue_overflow.inc();
+                    }
+                    Err(TrySendError::Closed(_)) => return,
+                }
+            }
+        }
+    }
+}
+
+async fn process_pods(
+    client: Client,
+    controller_pod_name: String,
+    mut rx: Receiver<ObjectReference>,
+    metrics: Metrics,
+) {
+    while let Some(object_ref) = rx.recv().await {
+        let namespace = object_ref.namespace.clone().unwrap_or_default();
+        let name = object_ref.name.clone().unwrap_or_default();
+        let pods = kube::Api::<Pod>::namespaced(client.clone(), &namespace);
+        let delete_res = pods.delete(&name, &Default::default()).await;
```

Comment:
Are the default params appropriate here?

https://docs.rs/kube/latest/kube/api/struct.DeleteParams.html

Specifically it seems like we may want to use a zero grace period.

Reply:
Good call :+1: